File: api/src/main/java/me/kvalbrus/multibans/api/DataProvider.java
package me.kvalbrus.multibans.api;

import java.sql.SQLException;
import java.util.List;
import java.util.UUID;
import me.kvalbrus.multibans.api.punishment.Punishment;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface DataProvider {

    String getName();

    void initialization() throws SQLException;

    void shutdown();

    void wipe();

    boolean createPunishment(@NotNull Punishment punishment);

    boolean updatePunishment(@NotNull Punishment punishment);

    boolean deletePunishment(@NotNull Punishment punishment);

    boolean hasPunishment(String id);

    @Nullable
    Punishment getPunishment(String id);

    @NotNull
    <T extends Punishment> List<T> getTargetHistory(UUID uuid);

    @NotNull
    <T extends Punishment> List<T> getTargetHistory(String name);

    @NotNull
    <T extends Punishment> List<T> getCreatorHistory(String creator);
}

File: api/src/main/kotlin/me/kvalbrus/multibans/api/DataProvider.kt
package me.kvalbrus.multibans.api

import me.kvalbrus.multibans.api.punishment.Punishment
import java.util.*

interface DataProvider {
    val name: String?

    @Throws(Exception::class) fun initialization()
    fun shutdown()
    @Throws(Exception::class) fun wipe()
    @Throws(Exception::class) fun createPunishment(punishment: Punishment): Boolean
    @Throws(Exception::class) fun createPlayerSession(session: Session)
    @Throws(Exception::class) fun savePunishment(punishment: Punishment): Boolean
    @Throws(Exception::class) fun deletePunishment(punishment: Punishment)
    @Throws(Exception::class) fun hasPunishment(id: String): Boolean
    @Throws(Exception::class) fun getPunishment(id: String): Punishment?
    @Throws(Exception::class) fun <T : Punishment?> getTargetHistory(uuid: UUID?): List<T>
    @Throws(Exception::class) fun <T : Punishment?> getTargetHistory(name: String?): List<T>
    @Throws(Exception::class) fun <T : Punishment?> getTargetHistoryByIp(ip: String): List<T>
    @Throws(Exception::class) fun <T : Punishment?> getCreatorHistory(creator: String?): List<T>
}